File: java/src/aoc2021/Day2.java
package aoc2021;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.concurrent.atomic.AtomicInteger;

public class Day2 {

    public static void main(String[] args) throws IOException {

        String sampleFilePath = "inputs/day2sample.txt";
        String inputFilePath = "inputs/day2input.txt";
        final AtomicInteger hVal = new AtomicInteger();
        final AtomicInteger dVal = new AtomicInteger();

        Files.readAllLines(Paths.get(inputFilePath)).stream().map(a -> a.split(" ")).parallel().forEach(
                line -> {
                    switch (line[0]) {
                        case "forward":
                            hVal.addAndGet(Integer.parseInt(line[1]));
                            break;
                        case "down":
                            dVal.addAndGet(Integer.parseInt(line[1]));
                            break;

                        case "up":
                            dVal.addAndGet(-Integer.parseInt(line[1]));
                            break;
                    }
                }
        );

        System.out.println(
                String.format("h: %d d: %d product: %d", hVal.get(), dVal.get(), hVal.get() * dVal.get())
        );

    }
}

File: scala/src/adventofcode2021/Day2.scala
package adventofcode2021

import scala.io.Source

object Day2 {

  def main(args: Array[String]): Unit = {
    val sampleFilePath = "inputs/day2sample.txt"
    val inputFilePath = "inputs/day2input.txt"
    var depth = 0
    var x = 0
    var aim = 0

    def part1Calculate(entry: Array[String]): Unit = {
      entry(0) match {
        case "forward" => x += entry(1).toInt
        case "down" => depth += entry(1).toInt
        case "up" => depth -= entry(1).toInt
      }
    }

    def part2Calculate(entry: Array[String]): Unit = {
      entry(0) match {
        case "forward" => { x += entry(1).toInt; depth += entry(1).toInt*aim}
        case "down" => aim += entry(1).toInt
        case "up" => aim -= entry(1).toInt
      }
    }

    Source.fromFile(inputFilePath)
      .getLines.toList.map(_.split(" ")).foreach(
      part2Calculate
    )

    println(s"h: ${x} depth: ${depth} product=${depth * x}")
  }



}
